_posts/2021-04-05-fmf-unified-testing.md

```diff
@@ -0,0 +1,156 @@
+  # https://github.com/packit/sandcastle/pull/92
+  # https://medium.com/the-node-js-collection/node-js-memory-management-in-container-environments-7eb8409a74e8
+  create-archive: make NODE_OPTIONS=--max-old-space-size=500 dist-gzip
+  # starter-kit.git has no release tags; your project can drop this once you have a release
+  get-current-version: make print-version
+jobs:
+  - job: tests
+    trigger: pull_request
+    metadata:
+      targets:
+      - fedora-all
+```
+
+This binds together the knowledge how to produce an upstream release tarball from your branch, where the spec file is, and on which Fedora releases to run tests in a PR. Packit will build the tarball (`create-archive`), build an SRPM with the spec file, build it in a temporary [COPR](https://copr.fedorainfracloud.org/), and use `tmt` to run your tests against these built RPMs. As a consumer, it really can't get much simpler than that!
+
+If you look on a [recent starter-kit PR](https://github.com/cockpit-project/starter-kit/pull/442), click on "View Details" to expand the tests, there you can see the four Packit runs.
+
+It is still not obvious how to get from such a [result link](http://artifacts.dev.testing-farm.io/f24bede3-995e-4f10-970f-dc849e950e3a/) to all artifacts (append `/results.xml` to find out the path to the `/work-allXXXXXX` directory, then open that), or getting some [live logs while a test is running](https://gitlab.com/testing-farm/general/-/issues/13), but that will surely be improved really soon.
+
+## Recent Fedora CI changes
+
+As mentioned above, Fedora's gating tests recently switched to run on the very same [testing-farm](https://testing-farm.gitlab.io/api/), so that Fedora gating and Packit tests run in the same environment and thus compare very well. It also supports FMF/TMT test metadata in addition to the old STI.
+
+This gets us *really* close to the goal of sharing the tests upstream and downstream. The only bit that's missing is that there is no clean way how to run tests that are contained in your upstream tarball -- right now, the packaging dist-git must have a top-level [FMF test plan](https://src.fedoraproject.org/rpms/cockpit/blob/rawhide/f/plans/upstream.fmf) like this:
+
+```yaml
+discover:
+  how: fmf
+  repository: https://github.com/cockpit-project/cockpit
+  # FIXME: get rid of the hardcoding: https://github.com/psss/tmt/issues/585
+  ref: "241"
+execute:
+  how: tmt
+```
+
+This checks out the tests from your upstream project git. This is awkward, as it requires accessing a remote git even though the tests are already in the source tarball; and it needs to know which exact tag to check out, so that the tests match your packaged release. This requires some [tmt design discussion](https://github.com/psss/tmt/issues/585), but for now we just [hacked our release scripts](https://github.com/cockpit-project/cockpituous/commit/2ef3f6c99912) to automatically bump the `ref:` in the test plan when it commits the new release to dist-git.
```

Comment:
Does this mean the Fedora gating tests will be guaranteed to only pass on a very recently built source tarball? If so, it would probably be good to state drawback somehow. Some of the readers may not connect the dots as to why you really this is broken.

Reply:
No, they will always pass -- that's exactly the reason why the `ref:` is a stable release tag instead of main/master. I (hope that I) clarified this a bit now.